addons/controllers/clusterbootstrap_handlers.go

```diff
@@ -76,6 +79,52 @@ func (r *ClusterBootstrapReconciler) ClusterBootstrapToClusters(o client.Object)
 	return []ctrl.Request{{NamespacedName: client.ObjectKeyFromObject(cluster)}}
 }
 
+// SecretsToClusters is the Map Function for watching Secrets that filters the events on
+// objects of type secret and returns requests for reconcile if required
+func (r *ClusterBootstrapReconciler) SecretsToClusters(o client.Object) []ctrl.Request {
+	secret, ok := o.(*corev1.Secret)
+	if !ok {
+		r.Log.Error(errors.New("invalid type"),
+			"Expected to receive Secret resource",
+			"actualType", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	log := r.Log.WithValues(constants.SecretNameLogKey, secret.Name)
+	log.V(4).Info("Mapping secrets to cluster")
+
+	// Here we filter based on two criteria
+	// 1. Secrets having the type ClusterBootstrapManagedSecret, OR
+	// 2. Secrets with ClusterNameLabel set
+	// For secrets that are cloned by ClusterBootstrap, we set the type and the first filter is used. We also
+	// set the ClusterNameLabel on these cloned secrets
+	// For other secrets such as those that we get from provider status (extensible provider model), the second
+	// filter is used because it is not possible to patch the Type of these secrets because the Type may be immutable
+	//
+	// Need to confirm: We can just use the second filter that is based on the cluster label for all secrets.
+	if secret.Type == constants.ClusterBootstrapManagedSecret {
+		for _, ownerRef := range o.GetOwnerReferences() {
+			if ownerRef.APIVersion == clusterv1beta1.GroupVersion.String() {
+				return []ctrl.Request{{NamespacedName: types.NamespacedName{Namespace: o.GetNamespace(), Name: ownerRef.Name}}}
+			}
+		}
+	}
+	cluster := &clusterv1beta1.Cluster{}
+	clusterName := ""
+	if secret.GetLabels() != nil {
+		clusterName = secret.GetLabels()[addontypes.ClusterNameLabel]
+		if clusterName != "" {
+			if err := r.Client.Get(r.context, client.ObjectKey{Namespace: secret.Namespace, Name: clusterName}, cluster); err != nil {
```

Comment:
Don't need to get the cluster, I would just return the objectkey. Reconcile loop will fetch anyway.